src/struct_mv/struct_matmult.c

```diff
@@ -1220,43 +1224,80 @@ hypre_StructMatmultCompute( hypre_StructMatmultData  *mmdata,
+   hypre_BoxArray            *Mdata_space    = hypre_StructMatrixDataSpace(M);
+   hypre_StTerm              *st_term; /* Pointer to stencil info for each term in a */
 
    /* Local variables */
-   hypre_Index            Mstart;      /* M's stencil location on the base index space */
-   hypre_Box             *loop_box;    /* boxloop extents on the base index space */
-   hypre_IndexRef         loop_start;  /* boxloop start index on the base index space */
-   hypre_IndexRef         loop_stride; /* boxloop stride on the base index space */
-   hypre_Index            loop_size;   /* boxloop size */
-   hypre_Index            Mstride;     /* data-map stride  (base index space) */
-   hypre_IndexRef         offsetref;   /* offset for constant coefficient stencil entries */
-   hypre_IndexRef         shift;       /* stencil shift from center for st_term */
-   hypre_IndexRef         stride;
+   hypre_Index                Mstart;      /* M's stencil location on the base index space */
+   hypre_Box                 *loop_box;    /* boxloop extents on the base index space */
+   hypre_IndexRef             loop_start;  /* boxloop start index on the base index space */
+   hypre_IndexRef             loop_stride; /* boxloop stride on the base index space */
+   hypre_Index                loop_size;   /* boxloop size */
+   hypre_Index                Mstride;     /* data-map stride  (base index space) */
+   hypre_IndexRef             offsetref;   /* offset for constant coefficient stencil entries */
+   hypre_IndexRef             shift;       /* stencil shift from center for st_term */
+   hypre_IndexRef             stride;
 
    /* Boxloop variables */
-   hypre_Index            fdstart,  cdstart,  Mdstart;  /* data starts */
-   hypre_Index            fdstride, cdstride, Mdstride; /* data strides */
-   hypre_Box             *fdbox,   *cdbox,   *Mdbox;    /* data boxes */
-   hypre_Index            tdstart;
+   hypre_Index                fdstart,  cdstart,  Mdstart;  /* data starts */
+   hypre_Index                fdstride, cdstride, Mdstride; /* data strides */
+   hypre_Box                 *fdbox,   *cdbox,   *Mdbox;    /* data boxes */
+   hypre_Index                tdstart;
 
    /* Indices */
-   HYPRE_Int              entry, Mentry;
-   HYPRE_Int              Mj, Mb;
-   HYPRE_Int              b, i, id, m, t;
+   HYPRE_Int                  entry, Mentry;
+   HYPRE_Int                  Mj, Mb;
+   HYPRE_Int                  b, i, id, m, t;
+
+#if defined(HYPRE_USING_GPU)
+   HYPRE_MemoryLocation       memory_location = hypre_StructMatrixMemoryLocation(matrices[0]);
+   HYPRE_ExecutionPolicy      exec_policy     = hypre_GetExecPolicy1(memory_location);
+#endif
 
    HYPRE_ANNOTATE_FUNC_BEGIN;
+   hypre_GpuProfilingPushRange("StructMatmultCompute");
+
+   /* Create work arrays on host memory for storing constant coefficients */
+   A_const = hypre_TAlloc(HYPRE_Complex*, nmatrices + 1, HYPRE_MEMORY_HOST);
```

Comment:
This is for future discussion, but I'm wondering if we should have a constant host array pointer in every matrix that would be a copy of the device data (when using GPUs) and add a HostConstData() macro to access it.  For CPU runs, the host data would just be a pointer to the device data, so no copying needed.  For GPU runs, we would need a SyncHostConstData() function (or something) that we would call first to ensure that the host data is correct.  Just a thought.  It might make for slightly cleaner code and maybe some reuse in other parts of the code.

Reply:
I like this idea, specially since it will make things cleaner.

Reply:
Okay, let's keep it in mind.